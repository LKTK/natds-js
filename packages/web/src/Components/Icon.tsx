import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { withTheme } from '@material-ui/styles';
import { IIconSizes } from '@naturacosmeticos/natds-styles';
import { iconNames } from '@naturacosmeticos/natds-icons';
import { IThemeWeb } from 'Themes';
import '@naturacosmeticos/natds-icons/dist/natds-icons.css';

export interface IIconProps {
  name: keyof typeof iconNames;
  size?: keyof IIconSizes;
  theme: IThemeWeb | unknown;
}

const Icon: FunctionComponent<IIconProps> = (props: IIconProps) => {
  const { name, theme, size } = props;
  const className = `natds-icons natds-icons-${String(name)}`;

  return (
    <IconComponent
      className={className}
      theme={theme}
      size={size}
    />
  );
};

export default withTheme(Icon);

const IconComponent = styled.i<Pick<IIconProps, 'size' | 'theme'>>`
  font-size: ${({ size, theme }) => {
    const { iconSizes = {} } = theme || {};
    const fontSize = size ? iconSizes[size] : iconSizes.standard;
    return `${fontSize}px`;
  }};
`;
