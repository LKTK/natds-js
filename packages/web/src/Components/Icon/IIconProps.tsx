import {iconNames} from '@naturacosmeticos/natds-icons';
import {IIconSizes} from '@naturacosmeticos/natds-styles';
import {IThemeWeb} from '../../Themes';

export interface IIconProps {
  className?: string;
  name: keyof typeof iconNames;
  size?: keyof IIconSizes;
  theme: IThemeWeb | unknown;
}
