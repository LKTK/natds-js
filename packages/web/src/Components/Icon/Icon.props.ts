import { iconNames } from "@naturacosmeticos/natds-icons";
import { IIconSizes } from "@naturacosmeticos/natds-styles";

export type IconName = keyof typeof iconNames;

export interface IIconProps {
  className?: string;

  /**
   * The icon name
   * @required
   */
  name: IconName;

  /**
   * The icon size
   *
   * @default "standard"
   */
  size?: keyof IIconSizes;
}
