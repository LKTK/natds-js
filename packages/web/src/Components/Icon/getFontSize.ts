import {IIconSizes, ITheme} from "@naturacosmeticos/natds-styles";

/**
 * @see https://material-ui.com/pt/styles/api/#createstyles-styles-styles
 */
export const getFontSize = ({ size = 'standard', theme }: {size?: keyof IIconSizes, theme: ITheme}) => {
  const { iconSizes = {} } = theme || {};
  const fontSize = iconSizes[String(size)];
  return `${fontSize}px`;
};
