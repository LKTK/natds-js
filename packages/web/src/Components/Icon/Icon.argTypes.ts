import { iconNames } from "@naturacosmeticos/natds-icons";
import { tokens } from "@naturacosmeticos/natds-styles";
import { IIconProps } from "./Icon.props";

export const argTypes : Record<keyof IIconProps, unknown> = {
  className: {
    control: {
      type: "text",
    },
  },
  name: {
    control: {
      options: Object.keys(iconNames),
      type: "select",
    },
    description: "The Icon name",
    type: {
      required: true,
    },
  },
  size: {
    control: {
      options: Object.keys(tokens.iconSizes),
      type: "select",
    },
    description: "The Icon size",
    type: {
      required: false,
    },
  },
};

export default argTypes;
