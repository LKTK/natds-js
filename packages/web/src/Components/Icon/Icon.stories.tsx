import * as React from "react";
import { Story } from "@storybook/react";
import { Icon, IIconProps } from "./Icon";
import argTypes from "./Icon.argTypes";

export default {
  argTypes,
  component: Icon,
  title: "Components/Icon",
};

const Template : Story<IIconProps> = (args) => <Icon {...args} />;

export const Playground : Story<IIconProps> = Template.bind({});
Playground.args = {};

export const StandardAddIcon : Story<IIconProps> = Template.bind({});
StandardAddIcon.args = {
  name: "outlined-action-add",
  size: "standard",
};

export const SmallIcon : Story<IIconProps> = Template.bind({});
SmallIcon.args = {
  ...Playground.args,
  size: "small",
};

export const SmallNaturaIcon : Story<IIconProps> = Template.bind({});
SmallNaturaIcon.args = {
  ...SmallIcon.args,
  name: "filled-brand-naturarosacea",
};

export const SmallCloseIcon : Story<IIconProps> = Template.bind({});
SmallCloseIcon.args = {
  ...SmallIcon.args,
  name: "outlined-navigation-close",
};

export const SmallAddIcon : Story<IIconProps> = Template.bind({});
SmallAddIcon.args = {
  ...SmallIcon.args,
  name: "outlined-action-add",
};

export const TinyIcon: Story<IIconProps> = Template.bind({});
TinyIcon.args = {
  ...Playground.args,
  size: "tiny",
};

export const TinyMenuIcon: Story<IIconProps> = Template.bind({});
TinyMenuIcon.args = {
  ...TinyIcon.args,
  name: "filled-navigation-menu",
};

export const OutlinedPersonIcon : Story<IIconProps> = Template.bind({});
OutlinedPersonIcon.args = {
  name: "outlined-social-person",
};

export const OutlinedFavoriteIcon : Story<IIconProps> = Template.bind({});
OutlinedFavoriteIcon.args = {
  name: "outlined-action-favorite",
};

export const OutlinedDeleteIcon : Story<IIconProps> = Template.bind({});
OutlinedDeleteIcon.args = {
  name: "outlined-action-delete",
};

export const OutlinedEmailIcon : Story<IIconProps> = Template.bind({});
OutlinedEmailIcon.args = {
  name: "outlined-communication-email",
};

export const Sizes : Story<IIconProps> = (args) => (
  <React.Fragment>
    <Icon {...args} size={"micro"} />
    <Icon {...args} size={"tiny"} />
    <Icon {...args} size={"small"} />
    <Icon {...args} size={"standard"} />
    <Icon {...args} size={"medium"} />
    <Icon {...args} size={"large"} />
    <Icon {...args} size={"largex"} />
    <Icon {...args} size={"largexx"} />
    <Icon {...args} size={"huge"} />
    <Icon {...args} size={"hugex"} />
    <Icon {...args} size={"hugexx"} />
  </React.Fragment>
);
Sizes.args = Playground.args;
