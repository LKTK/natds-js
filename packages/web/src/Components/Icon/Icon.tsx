import * as React from "react";
import MaterialIcon from "@material-ui/core/Icon";
import "@naturacosmeticos/natds-icons/dist/natds-icons.css";
import clsx from "clsx";
import { useStyles } from "./Icon.styles";
import { IIconProps } from "./Icon.props";

export { IIconProps } from "./Icon.props";

/**
 * ## Import
 *
 * ```
 * import { Icon } from '@naturacosmeticos/natds-web';
 * ```
 */
export const Icon = React.forwardRef<unknown, IIconProps>(
  (
    props: IIconProps,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ref: any,
  ) => {

    const {
      className,
      name = "filled-default-mockup",
      size = "standard",
      ...otherProps
    } = props;

    const classes = useStyles();
    const sizeClassNameKey = `${size}FontSize`;

    const classNames = clsx([
      "natds-icons",
      `natds-icons-${String(name)}`,
      className,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      classes[sizeClassNameKey],
    ]);

    return <MaterialIcon className={classNames} component={"i"} ref={ref} {...otherProps}>{name}</MaterialIcon>;
  },
);

Icon.displayName = "Icon";
export default Icon;
