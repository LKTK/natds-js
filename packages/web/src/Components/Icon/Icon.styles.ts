import makeStyles from "@material-ui/core/styles/makeStyles";
import { createStyles } from "@material-ui/core";
import { getFontSize } from "./getFontSize";
import { IThemeWeb } from "../../Themes";

// @todo adopt with base in prop: https://material-ui.com/pt/styles/basics/#adapting-based-on-props
const styles = (theme: IThemeWeb) => createStyles({
  microFontSize: {
    fontSize: getFontSize({ size: "micro", theme }),
  },
  tinyFontSize: {
    fontSize: getFontSize({ size: "tiny", theme }),
  },
  smallFontSize: {
    fontSize: getFontSize({ size: "small", theme }),
  },
  standardFontSize: {
    fontSize: getFontSize({ size: "standard", theme }),
  },
  mediumFontSize: {
    fontSize: getFontSize({ size: "medium", theme }),
  },
  largeFontSize: {
    fontSize: getFontSize({ size: "large", theme }),
  },

  /**
   * @deprecated since 0.19. It will be removed at 1.0.
   */
  largexFontSize: {
    fontSize: getFontSize({ size: "largeX", theme }),
  },
  largeXFontSize: {
    fontSize: getFontSize({ size: "largeX", theme }),
  },

  /**
   * @deprecated since 0.19. It will be removed at 1.0.
   */
  largexxFontSize: {
    fontSize: getFontSize({ size: "largeXX", theme }),
  },
  largeXXFontSize: {
    fontSize: getFontSize({ size: "largeXX", theme }),
  },
  hugeFontSize: {
    fontSize: getFontSize({ size: "huge", theme }),
  },

  /**
   * @deprecated since 0.19. It will be removed at 1.0.
   */
  hugexFontSize: {
    fontSize: getFontSize({ size: "hugeX", theme }),
  },
  hugeXFontSize: {
    fontSize: getFontSize({ size: "hugeX", theme }),
  },

  /**
   * @deprecated since 0.19. It will be removed at 1.0.
   */
  hugexxFontSize: {
    fontSize: getFontSize({ size: "hugeXX", theme }),
  },
  hugeXXFontSize: {
    fontSize: getFontSize({ size: "hugeXX", theme }),
  },
});

/**
 * @see https://material-ui.com/pt/guides/typescript/#using-createstyles-to-defeat-type-widening
 */
export const useStyles = makeStyles(styles, { name: "NatDSIcon" });

export default useStyles;
