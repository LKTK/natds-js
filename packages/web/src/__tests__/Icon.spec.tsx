import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { expect } from 'chai';

import Icon from '../Components/Icon';
import { tokens, IIconSizes, themes } from '@naturacosmeticos/natds-styles';

const iconSizes = Object.keys(tokens.iconSizes);

describe('Icon component', () => {
  test('should match snapshot - Any valid Icon with standard size', () => {
    const component = renderer.create(<Icon name="filled-action-add" />).toJSON();
    expect(component).matchSnapshot('Icon snapshot - valid icon with standard size');
  });

  iconSizes.forEach((size: string) => {
    test(`should match snapshot - Any valid Icon with ${size} size`, () => {
      const component = renderer.create(<Icon name="filled-action-add" size={size as keyof IIconSizes} theme={themes.natura.light} />).toJSON();
      expect(component).matchSnapshot(`Icon snapshot - valid icon with ${size} size`);
    });
  });

  test(`should match snapshot - Any valid Icon with empty theme`, () => {
    const component = renderer.create(<Icon name="filled-action-add" theme={{}} />).toJSON();
    expect(component).matchSnapshot(`Icon snapshot - valid Icon with empty theme`);
  });

  test(`should match snapshot - Any valid Icon with theme with undefined icon sizes`, () => {
    const component = renderer.create(<Icon name="filled-action-add" theme={{iconSizes: {}}} />).toJSON();
    expect(component).matchSnapshot(`Icon snapshot - valid Icon with theme with undefined icon sizes`);
  });

  test(`should match snapshot - Any valid Icon without theme`, () => {
    const component = renderer.create(<Icon name="filled-action-add" />).toJSON();
    expect(component).matchSnapshot(`Icon snapshot - valid Icon without theme`);
  });
});
