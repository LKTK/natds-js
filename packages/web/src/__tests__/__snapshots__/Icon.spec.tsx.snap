// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should match snapshot - Any valid Icon with huge size: Icon snapshot - valid icon with huge size 1`] = `
<i
  className="sc-bdVaJa dJsSCD natds-icons natds-icons-outlined-alert-cancel"
  size="huge"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with hugex size: Icon snapshot - valid icon with hugex size 1`] = `
<i
  className="sc-bdVaJa jyACMJ natds-icons natds-icons-outlined-alert-cancel"
  size="hugex"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with hugexx size: Icon snapshot - valid icon with hugexx size 1`] = `
<i
  className="sc-bdVaJa dGTqKp natds-icons natds-icons-outlined-alert-cancel"
  size="hugexx"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with hugexxx size: Icon snapshot - valid icon with hugexxx size 1`] = `
<i
  className="sc-bdVaJa eoKYvL natds-icons natds-icons-outlined-alert-cancel"
  size="hugexxx"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with large size: Icon snapshot - valid icon with large size 1`] = `
<i
  className="sc-bdVaJa jKoFUU natds-icons natds-icons-outlined-alert-cancel"
  size="large"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with larger size: Icon snapshot - valid icon with larger size 1`] = `
<i
  className="sc-bdVaJa dGRaGr natds-icons natds-icons-outlined-alert-cancel"
  size="larger"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with largest size: Icon snapshot - valid icon with largest size 1`] = `
<i
  className="sc-bdVaJa gBaqWx natds-icons natds-icons-outlined-alert-cancel"
  size="largest"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with medium size: Icon snapshot - valid icon with medium size 1`] = `
<i
  className="sc-bdVaJa ioRmCd natds-icons natds-icons-outlined-alert-cancel"
  size="medium"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with micro size: Icon snapshot - valid icon with micro size 1`] = `
<i
  className="sc-bdVaJa hQJWqK natds-icons natds-icons-outlined-alert-cancel"
  size="micro"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with small size: Icon snapshot - valid icon with small size 1`] = `
<i
  className="sc-bdVaJa eMHvyV natds-icons natds-icons-outlined-alert-cancel"
  size="small"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with standard size: Icon snapshot - valid icon with standard size 1`] = `
<i
  className="sc-bdVaJa jWbhAs natds-icons natds-icons-outlined-alert-cancel"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with standard size: Icon snapshot - valid icon with standard size 2`] = `
<i
  className="sc-bdVaJa dSgozt natds-icons natds-icons-outlined-alert-cancel"
  size="standard"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon with tiny size: Icon snapshot - valid icon with tiny size 1`] = `
<i
  className="sc-bdVaJa hQJWqK natds-icons natds-icons-outlined-alert-cancel"
  size="tiny"
/>
`;

exports[`Icon component should match snapshot - Any valid Icon without theme: Icon snapshot - valid Icon without theme 1`] = `
<i
  className="sc-bdVaJa jWbhAs natds-icons natds-icons-outlined-alert-cancel"
/>
`;
